File: tests/src/main/java/com/github/klee0kai/test/di/earthmirror/WestModule.java
package com.github.klee0kai.test.di.earthmirror;

import com.github.klee0kai.stone.annotations.module.Module;
import com.github.klee0kai.stone.annotations.module.Provide;
import com.github.klee0kai.test.mowgli.earth.*;

@Module
public abstract class WestModule {

    @Provide(cache = Provide.CacheType.Factory)
    public IRiver riverImpl() {
        return new River();
    }

    @Provide(cache = Provide.CacheType.Soft)
    public IMountain mountainImp() {
        return new Mountain();
    }

    @Provide(cache = Provide.CacheType.Soft)
    public abstract Cave cave();

    @Provide(cache = Provide.CacheType.Soft)
    public abstract Cave cave(Cave.CaveType type, Integer deep);

}

File: tests/src/commonMain/kotlin/com/github/klee0kai/test/di/earthmirror/WestModule.kt
package com.github.klee0kai.test.di.earthmirror

import com.github.klee0kai.stone.annotations.module.Module
import com.github.klee0kai.stone.annotations.module.Provide
import com.github.klee0kai.test.mowgli.earth.*
import com.github.klee0kai.test.mowgli.earth.Cave.CaveType

@Module
abstract class WestModule {
    @Provide(cache = Provide.CacheType.Factory)
    open fun riverImpl(): IRiver? {
        return River()
    }

    @Provide(cache = Provide.CacheType.Soft)
    open fun mountainImp(): IMountain? {
        return Mountain()
    }

    @Provide(cache = Provide.CacheType.Soft)
    abstract fun cave(): Cave?

    @Provide(cache = Provide.CacheType.Soft)
    abstract fun cave(type: CaveType?, deep: Int?): Cave?
}
